app/controllers/pipeline_viz_controller.rb

```diff
@@ -12,21 +12,47 @@ def show
     )
 
     if feature_allowed && pipeline_run
-      stage_info = {}
-      pipeline_run.pipeline_run_stages.each do |stage|
+      @stages = []
+      @edges = []
+      all_stage_dag_jsons = []
+      pipeline_run.pipeline_run_stages.each_with_index do |stage, stage_index|
         if stage.name != "Experimental" || current_user.admin?
-          stage_info[stage.name] = JSON.parse(stage.dag_json || "{}")
-          stage_info[stage.name][:job_status] = stage.job_status
+          stage_dag_json = JSON.parse(stage.dag_json || "{}")
+          stage_nodes = stage_node_scaffolding(stage_dag_json)
+
+          stage_edges = intra_stage_edges(stage_dag_json, stage_index)
+          stage_edges.each_with_index do |edge, edge_index|
+            if edge[:is_intra_stage]
+              # Offset index for correct value after concatination later
+              stage_nodes[edge[:from][:step_index]][:output_edges].push(edge_index + @edges.length)
+              stage_nodes[edge[:to][:step_index]][:input_edges].push(edge_index + @edges.length)
+            end
+          end
+          @edges.concat(stage_edges)
+
+          @stages.push(steps: stage_nodes,
+                       job_status: stage.job_status)
+          all_stage_dag_jsons.push(stage_dag_json)
+        end
+      end
+
+      between_stage_edges = inter_stage_edges(all_stage_dag_jsons, pipeline_run.pipeline_version)
+      between_stage_edges.each_with_index do |edge, edge_index|
+        # Offset index for correct value after concatination later
+        if edge[:from]
+          @stages[edge[:from][:stage_index]][:steps][edge[:from][:step_index]][:output_edges].push(edge_index + @edges.length)
```

Comment:
I think this could be more readable if you pull out `edge[:from][:stage_index]` as its own var.  So then this line could be something like

`@stages[edge_from][:steps][edge_from][:output_edges].push...`

Is it necessary to have the same key for to levels of your data structure here though? I know you're dealing with a lot of nested data here, but I'm wondering if it isn't redundant.

Reply:
I use the same key mainly so that (on the javascript side) the same accessor methods can be used regardless of input/output edges, but I will pull out the indices into separate variables in this part of the code to make it readable 👍